accepted/future-releases/nnbd/feature-specification.md

```diff
@@ -1236,18 +1248,92 @@ the `LEGACY_ERASURE` of the types.
+- **LEGACY_TYPE**(`X & T`) =
+  - This case should not occur, since intersection types are not permitted as
+    generic arguments.
+- **LEGACY_TYPE**(`C<T0, ..., Tn>`) = `C<R0, ..., Rn>*`
+  - where `Ri` is **LEGACY_TYPE**(`Ti`)
+- **LEGACY_TYPE**(`R Function<X extends B>(S)`) = `F*`
+  - where `F = R1 Function<X extends B1>(S1)`
+  - and `R1` = **LEGACY_TYPE**(`R`)
+  - and `B1` = **LEGACY_TYPE**(`B`)
+  - and `S1` = **LEGACY_TYPE**(`S`)
+
+Note that if `T` is a normal form type, then **LEGACY_TYPE**(`T`) is also a
+normal form type.
+
+#### Type literals
+
+Two constant type literals `T1` and `T2` canonicalize to the same object if they
+are equal using the definition of runtime type equality specified above.  The
+object `R` which serves as the canonical form for `T1` and `T2` is defined as
+follows. Let `S1` and `S2` be the types to which `T1` and `T2` correspond.  Note
+that by assumption, **NORM**(`S1`) and **NORM**(`S2`) are syntactically
+identical up to the identity of bound variables and the placement of legacy type
+modifiers.  We choose `R` arbitrarily as the type literal corresponding to one
+of **NORM**(`S1`) or **NORM**(`S2`) with all legacy type modifiers erased.
+
+The effect of this definition is to ensure that type literals which look
+identical in the source syntax but which may differ by the presence of legacy
+type modifiers are canonicalized consistently in the sense that any two type
+literals which would compare equal via the definition of runtime type equality
+given above will be canonicalized to the same type object.  The choice of what
+type object to canonicalize to is arbitrary in that placement of legacy
+modifiers in type literals is not otherwise observable in the language.  Any
+consistent choice of placement of legacy modifiers is hence an equally valid
+implementation.
+
+Note that the choice of canonicalization for type literals does not depend
+directly on whether sound or unsound null safety semantics are in use.  However,
+soundly null safe programs will never contain legacy types, and hence the
+erasure of legacy type modifiers is never necessary when running in sound mode.
+
+#### Constant instances
+
+In both sound and unsound null checking, canonicalization of generic object
+instances is performed relative to the normal forms of their generic type
+arguments.  That is, an instance of `C<T0>` canonicalizes to the same object as
+`C<T1>` if `T0` and `T1` have the same normal form (up to the identity of bound
+variables), and the objects are otherwise identical.
 
 With sound null checking, all generic const constructors and generic const
-literals are evaluated using the actual type arguments provided, whether legacy
-or non-legacy.  This ensures that with sound null checking, the final
-consistent semantics are obeyed.
+literals are evaluated using the actual type arguments provided.  This ensures
+that with sound null checking, the final consistent semantics are obeyed.
```

Comment:
In order to avoid that "2nd breaking change", we will need to specify that normalization is applied during canonicalization for a program with pure null safety. Surely we don't want to change `<int>[]` to `<int*>[]` when it occurs in a pure program, so we'll stop using `**LEGACY_TYPE**` and start using some other transformation when we switch to sound null checking. That other transformation might simply be normalization, because `**LEGACY_TYPE**(T) == T` whenever `T` is a type in a pure program (it doesn't contain any `*`, and it doesn't contain any type alias applications that denote a type that contains a `*`).